### PR TITLE
[SYCL][InvokeSimd] Add error for callee with struct return or argument

### PR DESCRIPTION
**Pull request by sarnex**
It doesn't work at runtime even if we make it through the compiler because of an IGC issue.

Once the IGC issue is fixed, we will need make some changes to the compiler and then lift this restriction.

### CODE REVIEW

**Comment by sarnex** [comment]
```

********************
Timed Out Tests (1):
  SYCL :: HostInteropTask/host-task-dependency4.cpp
```

not related

**Comment by sarnex** [comment]
I'll be on vacation until July 26th, I'll address any feedback then. Thanks!

**Review by turinevgeny — APPROVED** [review]

File: sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp [hunk 1]
```diff
@@ -364,6 +364,11 @@ constexpr bool has_ref_ret(Ret (*)(Args...)) {
   return std::is_reference_v<Ret>;
 }
 
+template <typename Ret, typename... Args>
+constexpr bool has_struct_ret(Ret (*)(Args...)) {
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp:368]: If I remember it correctly, the structures also cannot be passed to invoke_simd as an argument/parameter.
That is why attempt to pass local_accessor by value to invoke_simd doesn't work now.
Please add the checks for parameters as well.

Comment (sarnex) [reply sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp:373]: good point, thanks. i just pushed a commit adding the argument check as well as rebasing against current intel/llvm

File: sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp [hunk 2]
```diff
@@ -386,6 +396,14 @@ template <class Callable> constexpr void verify_callable() {
...
         !callable_has_ref_arg,
         "invoke_simd does not support callables with reference arguments");
+    constexpr bool callable_has_struct_ret = has_struct_ret(obj);
+    static_assert(
+        !callable_has_struct_ret,
+        "invoke_simd does not support callables returning structures");
+    constexpr bool callable_has_struct_arg = has_struct_arg(obj);
+    static_assert(
+        !callable_has_struct_arg,
+        "invoke_simd does not support callables with structure arguments");
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp:406]: invoke_simd SPEC allows structures, it is only the problem in GPU RT.
We need some way to enable struct args/rets for experiments, etc without modifying sources. Please add some macro to guard this check.
For example:
```c++
#ifndef __ESIMD_ENABLE_INVOKE_SIMD_STRUCTS
    constexpr bool callable_has_struct_ret = has_struct_ret(obj);
    static_assert(....);

#endif // end !__ESIMD_ENABLE_INVOKE_SIMD_STRUCTS
```

Comment (sarnex) [reply sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp:407]: just added this in latest commit, we already have `__INVOKE_SIMD_ENABLE_ALL_CALLABLES` so I based the name on that format, thanks

**Review by v-klochkov — APPROVED** [review]

File: sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp (hunk 2 above)
Comment (v-klochkov) [reply sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp:407]: Great. Thank you!

**Comment by v-klochkov** [comment]
HIP/AMD fails like this are unrelated to this PR:
what():  No device of requested type available. -1 (PI_ERROR_DEVICE_NOT_FOUND)

**Pull request merged**